storage/remote/wal_watcher.go

```diff
@@ -353,59 +291,125 @@ func (w *WALWatcher) watch(wl *wal.WAL, reader *wal.LiveReader, tail bool) error
-				level.Error(w.logger).Log("msg", "error reading to end of segment", "err", err)
+			err = w.readSegment(reader, segmentNum)
+
+			// Ignore errors reading to end of segment whilst replaying the WAL.
+			if !tail {
+				if err != nil && err != io.EOF {
+					level.Warn(w.logger).Log("msg", "ignoring error reading to end of segment, may have dropped data", "err", err)
+				} else if reader.Offset() != size {
+					level.Warn(w.logger).Log("msg", "expected to have read whole segment, may have dropped data", "segment", segmentNum, "read", reader.Offset(), "size", size)
+				}
+				return nil
+			}
+
+			// Otherwise, when we are tailing, non-EOFs are fatal.
+			if err != io.EOF {
+				return err
 			}
 
-			level.Info(w.logger).Log("msg", "a new segment exists, we should start reading it", "current", fmt.Sprintf("%08d", w.currentSegment), "new", fmt.Sprintf("%08d", last))
 			return nil
 
 		case <-readTicker.C:
-			if err := w.readSegment(reader); err != nil && err != io.EOF {
-				level.Error(w.logger).Log("err", err)
-				return err
-			}
-			if reader.TotalRead() >= size && !tail {
-				level.Info(w.logger).Log("msg", "done replaying segment", "segment", w.currentSegment, "size", size, "read", reader.TotalRead())
+			err = w.readSegment(reader, segmentNum)
+
+			// Ignore all errors reading to end of segment whilst replaying the WAL.
+			if !tail {
+				if err != nil && err != io.EOF {
+					level.Warn(w.logger).Log("msg", "ignoring error reading to end of segment, may have dropped data", "segment", segmentNum, "err", err)
+				} else if reader.Offset() != size {
+					level.Warn(w.logger).Log("msg", "expected to have read whole segment, may have dropped data", "segment", segmentNum, "read", reader.Offset(), "size", size)
+				}
 				return nil
 			}
+
+			// Otherwise, when we are tailing, non-EOFs are fatal.
+			if err != io.EOF {
+				return err
+			}
 		}
 	}
 }
 
-func (w *WALWatcher) readSegment(r *wal.LiveReader) error {
-	for r.Next() && !isClosed(w.quit) {
-		err := w.decodeRecord(r.Record())
+func (w *WALWatcher) garbageCollectSeries(segmentNum int) error {
+	dir, _, err := tsdb.LastCheckpoint(w.walDir)
+	if err != nil && err != tsdb.ErrNotFound {
+		return errors.Wrap(err, "tsdb.LastCheckpoint")
+	}
 
-		// Intentionally skip over record decode errors.
-		if err != nil {
-			level.Error(w.logger).Log("err", err)
+	if dir == "" || dir == w.lastCheckpoint {
+		return nil
+	}
+	w.lastCheckpoint = dir
+
+	index, err := checkpointNum(dir)
+	if err != nil {
+		return errors.Wrap(err, "error parsing checkpoint filename")
+	}
+
+	if index >= segmentNum {
+		level.Debug(w.logger).Log("msg", "current segment is behind the checkpoint, skipping reading of checkpoint", "current", fmt.Sprintf("%08d", segmentNum), "checkpoint", dir)
+		return nil
+	}
+
+	level.Debug(w.logger).Log("msg", "new checkpoint detected", "new", dir, "currentSegment", segmentNum)
+
+	// This potentially takes a long time, should we run it in another go routine?
```

Comment:
I left this comment a while ago; do we want to do this?

Reply:
Nah, it doesn't take that long.  Will remove comment.